File: 백준/Gold/15486. 퇴사 2/퇴사 2.py
N=int(input())
T = [0 for _ in range(1500010)]
P = [0 for _ in range(1500010)]
for i in range(1,N+1):
    t,p = map(int,input().split())
    T[i]=t
    P[i]=p

dp=[0 for _ in range(N+10)]
prev=0 #?????????
for i in range(1,N+2):
    prev = max(prev, dp[i])
    if i+T[i]>N+1: continue
    dp[i+T[i]] = max(prev+P[i],dp[i+T[i]])
# print(dp)
print(max(dp))

File: 백준/Gold/15486. 퇴사 2/퇴사 2.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] T = new int[N + 10];
        int[] P = new int[N + 10];
        for (int i = 0; i < N; i++) {
            String[] s = br.readLine().split(" ");
            int ti = Integer.parseInt(s[0]);
            int pi = Integer.parseInt(s[1]);
            T[i] = ti;
            P[i] = pi;
        }

        int[] dp = new int[N + 2];
        int curmax = 0;
        for (int i = 0; i < N + 2; i++) {
            curmax = Math.max(curmax, dp[i]);
            // System.out.println(Arrays.toString(arr[i]));
            if (i + T[i] > N + 1)
                continue;
            dp[i + T[i]] = Math.max(dp[i + T[i]], curmax + P[i]);
        }
        // Arrays.sort(dp);
        // System.out.println(Arrays.toString(dp));
        System.out.println(dp[N]);
    }
}